### ISSUE TITLE
Subsequent calls to multiple services compose files randomly makes Cloudformation to update services

### ISSUE THREAD

**Issue by flaviostutz** [docker/compose-cli#1427]
When doing subsequent calls for deploying a non changing docker-compose.yml file, sometimes Cloudformation redeploys some services because it identifies a change in cloudformation file, which makes the deploy much longer.

1. Create a docker-compose.yml file with two services
2. Deploy the compose file to ECS using compose
3. Deploy the same compose file again using compose
4. Check if a changeset was identified. It not, redeploy again until you see the changes. Files with more than 4 services almost everytime causes changes.

**Describe the results you expected:**
I expected those untouched services not to be redeployed, because it takes a lot of time.

**Insights:**
Investigating further, I found that the generated CloudFormation file randomly changes the order of the services, causing CF to create an active changeset. Maybe forcing a sorting in services before template generation would fix this.

**Output of `docker version`:**

```
Client: Docker Engine - Community
 Cloud integration: 1.0.7
 Version:           20.10.2
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        2291f61
 Built:             Mon Dec 28 16:12:42 2020
 OS/Arch:           darwin/amd64
 Context:           ecs3
 Experimental:      true

Server: Docker Engine - Community
 Engine:
  Version:          20.10.2
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       8891c58
  Built:            Mon Dec 28 16:15:28 2020
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.3
  GitCommit:        269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc:
  Version:          1.0.0-rc92
  GitCommit:        ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0
```

**Output of `docker info`:**

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  app: Docker App (Docker Inc., v0.9.1-beta3)
  buildx: Build with BuildKit (Docker Inc., v0.5.1-docker)
  scan: Docker Scan (Docker Inc., v0.5.0)

Server:
 Containers: 3
  Running: 0
  Paused: 0
  Stopped: 3
 Images: 2082
 Server Version: 20.10.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Native Overlay Diff: true
 Logging Driver: json-file
 Cgroup Driver: cgroupfs
 Cgroup Version: 1
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 io.containerd.runtime.v1.linux runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 269548fa27e0089a8b8278fc4fc781d7f65a939b
 runc version: ff819c7e9184c13b7c2607fe6c30ae19403a7aff
 init version: de40ad0
 Security Options:
  seccomp
   Profile: default
 Kernel Version: 4.19.121-linuxkit
 Operating System: Docker Desktop
 OSType: linux
 Architecture: x86_64
 CPUs: 6
 Total Memory: 7.778GiB
 Name: docker-desktop
 ID: LB4S:TNW6:OYPM:RCDP:6MCE:5VYU:HQR2:I5PW:NAYA:EVKE:OALR:3ENI
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 HTTP Proxy: gateway.docker.internal:3128
 HTTPS Proxy: gateway.docker.internal:3129
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false```

**Additional environment details (AWS ECS, Azure ACI, local, etc.):**
ECS

**Comment by ndeloof** [comment]
Thanks for reporting this issue and to propose a fix! 👍 

So CloudFormation is resource order-sensitive... ಠ_ಠ 

Anyway, I wonder this will actually solve this issue: goformation template.Resource is actually a map (https://github.com/awslabs/goformation/blob/8ccf34c68b6e1302b47b84c961b92472abb1bde6/cloudformation/template.go#L55) so - by Golang design - is unpredictable in key,values ordering. Also other resources should be affected, shouldn't they?

It seems to me to get this fixed we need to ensure it get marshalled in key order.

**Comment by flaviostutz** [comment]
You are right about the hashmap ordering issue, but in some way this was working (I tested a lot). This fix is part of that "old" PR with lots of things :) that we discussed to break into smaller PRs so maybe there is something missing in the PR.
I will take a better look.

**Comment by flaviostutz** [comment]
I realized Golang marshal is ordering map keys by default, so resource declaration order is not a problem. I've tested current main branch with a bunch of different scenarios and "convert" always returns the same "yml", so there is no bug currently.

I had this problem when I was implementing something causing two services to add things to a list that was rendered on a third service, so the order that we processed the services changed the list order and the "third" service was being updated randomly (because marshal follows list ordering). 

We may close this issue/PR, but indeed, when template changes order it redeploys the entire service/task.
